src/Microsoft.Data.Analysis/DataFrame.IO.cs

```diff
@@ -214,87 +193,189 @@ private static DataFrameColumn CreateColumn(Type kind, string[] columnNames, int
+            {
+                line = enumerator.Current;
+                var spl = line.Split(separator);
+                if (header && rowline == 0)
                 {
-                    var spl = line.Split(separator);
-                    if (header && rowline == 0)
-                    {
-                        // Skips.
-                    }
-                    else
-                    {
-                        ret.Append(spl, inPlace: true);
-                    }
-                    ++rowline;
-                    line = streamReader.ReadLine();
+                    // Skips.
+                }
+                else
+                {
+                    ret.Append(spl, inPlace: true);
                 }
+                ++rowline;
+            }
 
-                if (addIndexColumn)
+            if (addIndexColumn)
+            {
+                PrimitiveDataFrameColumn<int> indexColumn = new PrimitiveDataFrameColumn<int>("IndexColumn", columns[0].Length);
+                for (int i = 0; i < columns[0].Length; i++)
                 {
-                    PrimitiveDataFrameColumn<int> indexColumn = new PrimitiveDataFrameColumn<int>("IndexColumn", columns[0].Length);
-                    for (int i = 0; i < columns[0].Length; i++)
-                    {
-                        indexColumn[i] = i;
-                    }
-                    columns.Insert(0, indexColumn);
+                    indexColumn[i] = i;
                 }
-                return ret;
+                columns.Insert(0, indexColumn);
+            }
+            return ret;
+        }
+
+        private struct CsvLines : IEnumerable<string>
+        {
+            private CsvLineEnumerator enumerator;
+            public CsvLines(CsvLineEnumerator csvLineEnumerator)
+            {
+                enumerator = csvLineEnumerator;
+            }
+
+            public IEnumerator<string> GetEnumerator() => enumerator;
+
+            IEnumerator IEnumerable.GetEnumerator() => enumerator;
+        }
+
+        private struct CsvLineEnumerator : IEnumerator<string>
+        {
+            private StreamReader streamReader;
+            private string currentLine;
+            private long streamStartPosition;
+            public CsvLineEnumerator(StreamReader csvStream)
+            {
+                streamStartPosition = csvStream.BaseStream.Position;
+                streamReader = csvStream;
+                currentLine = null;
+            }
+
+            public string Current => currentLine;
+
+            object IEnumerator.Current => currentLine;
+
+            public void Dispose()
+            {
+                throw new NotImplementedException();
+            }
+
+            public bool MoveNext()
+            {
+                currentLine = streamReader.ReadLine();
+                return currentLine != null;
+            }
+
+            public void Reset()
```

Comment:
Do we need to support Reset()? Who calls it? You can save a `long` field if it wasn't implemented.

Reply:
Ah, nevermind. I see where we call it now.